winsup/utils/getprocaddr.c

```diff
@@ -249,32 +192,119 @@ main (int argc, char **argv)
       fprintf (stderr, "Could not register Ctrl handler\n");
       return 1;
     }
+
   if (!GenerateConsoleCtrlEvent (CTRL_BREAK_EVENT, 0))
     {
       fprintf (stderr, "Could not simulate Ctrl+Break\n");
       return 1;
     }
 
-  CtrlEvent = CreateEvent (NULL,              // default security attributes
-                           TRUE,              // manual-reset event
-                           FALSE,             // initial state is nonsignaled
-                           TEXT ("CtrlEvent") // object name
-  );
-  if (CtrlEvent == NULL)
+  if (WaitForSingleObject (CtrlEvent, 10000 /* 10 seconds*/) != WAIT_OBJECT_0)
     {
-      fprintf (stderr, "CreateEvent failed (%ld)\n", GetLastError ());
+      fprintf (stderr, "WaitForSingleObject failed (%ld)\n", GetLastError ());
       return 1;
     }
+  return 0;
+}
 
-  if (WaitForSingleObject (CtrlEvent, 10000 /* 10 seconds*/) != WAIT_OBJECT_0)
+static void *
+get_proc_addr (const char * module_name, const char * function_name)
+{
+      HMODULE module = GetModuleHandle (module_name);
+      if (!module)
+        return NULL;
+      return (void *)GetProcAddress (module, function_name);
+}
+
+int
+main (int argc, char **argv)
+{
+  char *end;
+  void *address;
+  BOOL is_ctrl_routine;
+  DWORD thread_return = 0;
+
+  if (argc == 4)
     {
-      fprintf (stderr, "WaitForSingleObject failed (%ld)\n", GetLastError ());
+      exit_code = atoi (argv[2]);
+      pid = strtoul (argv[3], NULL, 0);
+    }
+  else if (argc == 2)
+    {
+      pid = 0;
+    }
+  else
+    {
+      fprintf (stderr, "Need a function name, exit code and pid\n"
+               "Or needs a function name.\n");
+      return 1;
+    }
+
+  is_ctrl_routine = strcmp (argv[1], "CtrlRoutine") == 0;
+  address = get_proc_addr ("kernel32", argv[1]);
+  if (is_ctrl_routine && !address)
+    {
+      /* CtrlRoutine is undocumented, and has been seen in both
+       * kernel32 and kernelbase
+       */
+      address = get_proc_addr ("kernelbase", argv[1]);
+      if (!address)
+        return find_ctrl_routine_the_hard_way ();
+    }
+
+  if (!address)
+    {
+      fprintf (stderr, "Could not get proc address\n");
+      return 1;
+    }
+
+  if (pid == 0)
+    {
+      printf ("%p\n", address);
+      fflush (stdout);
+      return 0;
+    }
+  HANDLE h = OpenProcess (PROCESS_CREATE_THREAD |
+                          PROCESS_QUERY_INFORMATION | PROCESS_VM_OPERATION |
+                          PROCESS_VM_WRITE | PROCESS_VM_READ, FALSE, pid);
+  if (h == NULL)
+    {
+      fprintf (stderr, "OpenProcess failed: %ld\n", GetLastError ());
+      return 1;
+    }
+  /* Inject the remote thread */
+  if (inject_remote_thread_into_process (h, (LPTHREAD_START_ROUTINE)address,
+                                         exit_code, &thread_return) < 0)
+    {
+      fprintf (stderr, "Could not inject thread into process %lu\n", pid);
       return 1;
     }
+
+  if (is_ctrl_routine && thread_return)
+    {
+      fprintf (stderr,
```

Comment:
I don't really know whether it makes sense to have this `fprintf()` (and the likewise new one in `ctrl_handler()`). In the regular case, it will go to a hidden console. I guess it does not hurt... But does it benefit anyone?

Reply:
@naveen521kk ? https://github.com/msys2/msys2-runtime/pull/52#discussion_r656997079

Reply:
I guess I can live with it as-is ;-)

Reply:
I would still like to hear @naveen521kk's rationale, but it is nice to see when testing.

Reply:
While testing the executables through bash I could see it didn't print to hidden console, instead the output was in original console. That's why I wanted it. But yes, this will not work with cmd.